diesel/src/expression/helper_types.rs

```diff
@@ -13,7 +13,7 @@ pub type SqlTypeOf<Expr> = <Expr as Expression>::SqlType;
 pub type AsExpr<Item, TargetExpr> = AsExprOf<Item, SqlTypeOf<TargetExpr>>;
 
 /// The type of `Item` when converted to an expression of `Type`
-pub type AsExprOf<Item, Type> = <Item as AsExpression<Type>>::Expression;
+pub type AsExprOf<Item, Type> = <Item as AsExpressionOrArrayComparison<Type>>::ExpressionOrArrayCmp;
```

Comment:
This should be done as separate type definition. The existing `AsExprOf` impl remains useful on its own.

Reply:
Added a new type.